### ISSUE TITLE
TLD Prefix Rewrite

### ISSUE THREAD

**Issue by ikreymer** [hypothesis/via#23]
I've been experimenting with supporting the addition of top-level domain as a prefix in pywb.
For example, instead of `https://via.hypothes.is/example.com/` the url would be `https://example-com-via.hypothes.is/example.com/`

For sub-domains, only the TLD would be used, eg. `https://example-com-via.hypothes.is/a.b.c.example.com/path.html`

A 302 redirect would send the user from the base via to the correct subdomained version, eg. 
https://via.hypothes.is/example.com/` -> https://example-com-via.hypothes.is/example.com/`

The primary reason for this is to separate out cookies so that cookies from one domain are only accessible to cookies from the same TLD. eg, `example-A.com` cookies would be under `example-A-com-via.hypothes.is` while `example-B.com` cookies would be under `example-B-com-via.hypothes.is`

HTTPS wildcard only supports one level of depth, so can not have infinite subdomains with HTTPS.

A secondary benefit is that it may allow for pages to load faster, as browsers limit number of connections to a single host, but with multiple subdomains, these limits are raised.

Any thoughts on this, or any issues that you could foresee?

**Comment by tilgovi** [comment]
Cookies are not generally shared among subdomains. This would cause them to be shared. That could cause problems.

**Comment by ikreymer** [comment]
True, although, currently all cookies are set on `via.hypothes.is`. This will at least split them up by top user domain. It is not too uncommon for `example.com` to share cookies with `subdomain.example.com` with `domain=*.example.com`, in this case it would just be `domain=example-com-via.hypothes.is`

The main concern really is security, as currently cookies can be shared across all domain.

**Comment by ikreymer** [comment]
The main downside i see is the propagation of longer urls, users would see `https://example-com-via.hypothes.is/example.com/` in the location bar instead of `https://via.hypothes.is/example.com/`
(Only way around that is to use frames mode)

**Comment by tilgovi** [comment]
We don't have to proxy cookies. We could just strip them. That's a degraded
experience in some places, but it might be better than the security issues
of sharing one cookie domain. If this isn't going to solve that, I wonder
if it's worthwhile.

We could turn SSL off, and use any number of subdomains of any depth, as an
alternative.

**Comment by ikreymer** [comment]
There's also JavaScript-created cookies, eg. `document.cookie = ...` which now gets rewritten as well to go into the current domain. Disabling these will probably affect certain sites, such as NYTimes...

Another option is to use the full domain in the prefix, eg. `subdomain-A-example-com`, `subdomain-B-example-com`.. This will split everything by domain, however, will break cookies that have a
`domain=.example.com` and require sharing of cookies across subdomains... I don't know how often that is.. This could be configurable, where the full domain is the default, but only top-level is used in certain cases. That already is significantly better than current situation.

Disabling HTTPS seems like the more dangerous option overall, and something I'd like to avoid in other use cases.
